### PR TITLE
PreValidateTrader

### PR DESCRIPTION
**Pull request by ChrisCookOC**

### CODE REVIEW

File: app/uk/gov/hmrc/excisemovementcontrolsystemapi/models/preValidateTrader/response/ExciseTraderErrorResponse.scala
Comment (mpbostock) [inline app/uk/gov/hmrc/excisemovementcontrolsystemapi/models/preValidateTrader/response/ExciseTraderErrorResponse.scala:1]: All these response objects could just be condensed to one where we either have valid details set or error details set, they're all optional anyway so can we not just have a single response that covers all scenarios?

File: it/uk/gov/hmrc/excisemovementcontrolsystemapi/PreValidateTraderControllerItSpec.scala [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+
+      postRequest(Json.toJson("{'blah': 'badJson'}")).status mustBe BAD_REQUEST
+    }
+
+    "return Unsupported Media Type (415)" in {
+      withAuthorizedTrader(authErn)
+      postRequest(Json.toJson(""), contentType = """application/xml""").status mustBe UNSUPPORTED_MEDIA_TYPE
+    }
+
+    "return bad request (400) when body is not xml" in {
```
Comment (mpbostock) [inline it/uk/gov/hmrc/excisemovementcontrolsystemapi/PreValidateTraderControllerItSpec.scala:137]: The body should be JSON, this test shouldn't be needed, but more worryingly, why does it pass?

File: resources/public/api/conf/1.0/application.yaml [hunk 2]
```diff
@@ -665,19 +688,140 @@ components:
...
+      enum: [ 'UK Record', 'EU Trader', 'EU Warehouse', 'EU Temporary Authorisation' ]
+      example: 'UK Record'
+    ValidateProductAuthorisationRequest:
+      type: object
+      required:
+        - product
+      properties:
+        product:
+          $ref: '#/components/schemas/ProductRequest'
+    ExciseTraderValidationSuccessResponse:
```
Comment (mpbostock) [inline resources/public/api/conf/1.0/application.yaml:728]: Could we simplify the responses to just one response with optional fields and the relevant examples for success and failure just fill in the relevant fields?

File: test/uk/gov/hmrc/excisemovementcontrolsystemapi/connector/util/ExtractPreValidateTraderResponseSpec.scala [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+class ExtractPreValidateTraderResponseSpec extends PlaySpec with EitherValues {
+
+  private val validResponse = getPreValidateTraderSuccessResponse
+  private val businessError = getPreValidateTraderErrorResponse
+
+  "extractResponse" should {
+
+    "return the PreValidateTraderResponse if Success returned by EIS" in {
+
+      val response = ExtractPreValidateTraderResponse.extractResponse(HttpResponse(200, Json.toJson(validResponse).toString()))
```
Comment (mpbostock) [inline test/uk/gov/hmrc/excisemovementcontrolsystemapi/connector/util/ExtractPreValidateTraderResponseSpec.scala:38]: If we simplified the response structure, we wouldn't need this class of this test

File: test/uk/gov/hmrc/excisemovementcontrolsystemapi/controllers/PreValidateTraderControllerSpec.scala [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+      when(connector.submitMessage(any, any)(any))
+        .thenReturn(Future.successful(Left(NotFound("not found"))))
+
+      val result = createWithSuccessfulAuth.submit(request)
+
+      status(result) mustBe NOT_FOUND
+    }
+
+    "a validation parser error" when {
+      "xml cannot be parsed" in {
```
Comment (mpbostock) [inline test/uk/gov/hmrc/excisemovementcontrolsystemapi/controllers/PreValidateTraderControllerSpec.scala:93]: Why are we testing with failed xml if the body is JSON?

File: test/uk/gov/hmrc/excisemovementcontrolsystemapi/controllers/PreValidateTraderControllerSpec.scala [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+
+  private def createWithAuthActionFailure =
+    new PreValidateTraderController(
+      FakeFailingAuthentication,
+      FakeSuccessJsonParser,
+      connector,
+      cc
+    )
+
+  private def createWithFailingXmlParserAction =
```
Comment (mpbostock) [inline test/uk/gov/hmrc/excisemovementcontrolsystemapi/controllers/PreValidateTraderControllerSpec.scala:118]: Ah, relating to the previous comment, I see that createWithFailingXmlparserAction is a typo and it's actually checking failed JSON parsing

File: app/uk/gov/hmrc/excisemovementcontrolsystemapi/models/preValidateTrader/response/ExciseTraderErrorResponse.scala
Comment (ChrisCookOC) [reply app/uk/gov/hmrc/excisemovementcontrolsystemapi/models/preValidateTrader/response/ExciseTraderErrorResponse.scala:1]: Done

File: resources/public/api/conf/1.0/application.yaml (hunk 2 above)
Comment (ChrisCookOC) [reply resources/public/api/conf/1.0/application.yaml]: Tried this but the resulting schema looked too confusing - wasn't clear when each item was used. So left as is

File: test/uk/gov/hmrc/excisemovementcontrolsystemapi/connector/util/ExtractPreValidateTraderResponseSpec.scala (hunk 3 above)
Comment (ChrisCookOC) [reply test/uk/gov/hmrc/excisemovementcontrolsystemapi/connector/util/ExtractPreValidateTraderResponseSpec.scala]: Removed

File: test/uk/gov/hmrc/excisemovementcontrolsystemapi/controllers/PreValidateTraderControllerSpec.scala (hunk 5 above)
Comment (ChrisCookOC) [reply test/uk/gov/hmrc/excisemovementcontrolsystemapi/controllers/PreValidateTraderControllerSpec.scala]: Renamed

**Review by mpbostock — APPROVED** [review]

**Pull request merged**